server/src/main/java/keywhiz/service/daos/SecretDAO.java

```diff
@@ -385,6 +403,48 @@ public void deleteSecretsByName(String name) {
         .deleteSecretSeriesByName(name);
   }
 
+  /**
+   * Counts the number of secrets deleted before the specified cutoff.
+   *
+   * @param deleteBefore the cutoff date; secrets deleted before this date will be counted
+   */
+  public int countSecretsDeletedBeforeDate(DateTime deleteBefore) {
+    checkArgument(deleteBefore != null);
+
+    // identify the secrets deleted before this date
+    return secretSeriesDAOFactory.using(dslContext.configuration())
+        .getIdsForSecretSeriesDeletedBeforeDate(deleteBefore)
+        .size();
+  }
+
+  /**
+   * Permanently removes the series and all secrets-contents records ("versions") which were deleted
+   * before the given date.
+   *
+   * Unlike the "delete" endpoints above, THIS REMOVAL IS PERMANENT and cannot be undone by editing
+   * the database to restore the "current" entries.
+   *
+   * @param deleteBefore the cutoff date; secrets deleted before this date will be deleted
+   */
+  public void dangerPermanentlyRemoveSecretsDeletedBeforeDate(DateTime deleteBefore) {
+    checkArgument(deleteBefore != null);
+    SecretSeriesDAO secretSeriesDAO = secretSeriesDAOFactory.using(dslContext.configuration());
+    SecretContentDAO secretContentDAO = secretContentDAOFactory.using(dslContext.configuration());
+
+    // identify the secrets deleted before this date
+    List<Long> ids = secretSeriesDAO.getIdsForSecretSeriesDeletedBeforeDate(deleteBefore);
+
+    // batch the list of secrets to be removed, to reduce load on the database
+    List<List<Long>> partitionedIds = Lists.partition(ids, MAX_ROWS_REMOVED_PER_TRANSACTION);
```

Comment:
It might be nice to have a sleep between batches, so we can tune how much load this incurs

Reply:
Good idea--added.